File: Algorithm_Mentoring/25-2/python/7주차/가장_긴_증가하는_부분_수열.py
import sys
sys.setrecursionlimit(10**6)
input = sys.stdin.readline

MAX = 1010
n = int(input())
a = [0] + list(map(int, input().split()))
dp = [-1] * MAX

def solve(cur):
    if dp[cur] != -1:
        return dp[cur]
    dp[cur] = 0

    for i in range(cur):
        # 이전의 값이 크거나 같으면 증가하는 부분 수열이 아님
        if a[i] >= a[cur]:
            continue
        # 이전의 수열 길이에 +1
        dp[cur] = max(dp[cur], solve(i) + 1)

    return dp[cur]

result = 0
dp[0] = 0

# 1 ~ n의 DP값 중 최댓값이 정답
for i in range(1, n + 1):
    result = max(result, solve(i))

print(result)

File: Algorithm_Mentoring/25-2/cpp/7주차/가장_긴_증가하는_부분_수열.cpp
#include <iostream>
using namespace std;
using ll = long long;

const ll MAX = 1010;
ll n, a[MAX], dp[MAX];

ll solve(ll cur){
    ll& ret = dp[cur];
    if(ret != -1) return ret; ret = 0;

    for(int i = 0;i < cur;i++){
        // 이전의 값이 크거나 같으면 증가하는 부분 수열이 아님
        if(a[i] >= a[cur]) continue;

        // 이전의 수열 길이에 +1
        ret = max(ret, solve(i) + 1);
    }

    return ret;
}

int main(){
    ios::sync_with_stdio(0); // fastio
    cin.tie(0), cout.tie(0); // fastio
    
    cin >> n;
    for(int i = 1;i <= n;i++) cin >> a[i];
    for(int i = 0;i < MAX;i++) dp[i] = -1;

    ll result = 0; dp[0] = 0;
    //1 ~ n의 DP값 중 최댓값이 정답
    for(int i = 1;i <= n;i++) result = max(result, solve(i));
    cout << result;

    return 0;
}

